mantidproject/mantid
# Improvements to summation in Q in RRO

This PR adds some fixes and improvements to the summation in Q functionality in `ReflectometryReductionOne`. See the linked issues for details. 

Fixes #20120 and fixes #20123 

**To test:**

1. Run the following script. You should get the plot as shown below.
2. Edit the first line to set `diagnostics=1` and re-run the script. You should get lots of additional interim workspaces output, as shown below.
```
diagnostics=0

pixelindex=403
pix_start=389
pix_end=418

def plotResults(ws,suffix,plot_title,xmin,xmax,ymin,ymax):
    '''Plot the LS, QSB and QSD workspaces with the given suffix'''
    lines = plot(['LS'+ws+suffix,'QSB'+ws+suffix,'QSD'+ws+suffix],  0, tool='plot_spectrum')
    fig = lines[0].figure()
    fig.suptitle("IvsQ")
    ax = fig.axes()[0]
    ax.set_xscale('log')
    ax.set_yscale('log')
    if title != None:
        title(plot_title)
    if xmin != None and xmax != None:
        ax.set_xlim(xmin,xmax)
    if ymin != None and ymax != None:
        ax.set_ylim(ymin,ymax)

def plotResultsQ(ws,title=None,xmin=None,xmax=None,ymin=None,ymax=None):
    '''Plot the binned Q workspaces'''
    plotResults(ws,'_binned',title,xmin,xmax,ymin,ymax)

def plotResultsQUnbinned(ws,title=None,xmin=None,xmax=None,ymin=None,ymax=None):
    '''Plot the unbinned Q workspaces'''
    plotResults(ws,'_refQ',title,xmin,xmax,ymin,ymax)
    
def plotResultsLam(ws,title=None,xmin=None,xmax=None,ymin=None,ymax=None):
    '''Plot the unbinned lambda workspaces'''
    plotResults(ws,'_refL',title,xmin,xmax,ymin,ymax)

def perfectreduce(runno=None, outputws="test", processing="397-415", resolution=0.01,xmin=None,xmax=None,ymin=None,ymax=None):
    #Do lambda summed reduction
    ReflectometryReductionOneAuto(InputWorkspace=str(outputws), SummationType='SumInLambda', ReductionType='Normal', AnalysisMode='MultiDetectorAnalysis', ProcessingInstructions=processing, FirstTransmissionRun="", CorrectionAlgorithm='None', MomentumTransferStep=resolution,  OutputWorkspaceBinned='LS'+str(outputws)+'_binned', OutputWorkspace='LS'+str(outputws)+'_refQ', OutputWorkspaceWavelength='LS'+str(outputws)+'_refL',StrictSpectrumChecking=False, WavelengthMax=13.5,Diagnostics=diagnostics)
    #Q summed for divergent beam
    ReflectometryReductionOneAuto(InputWorkspace=str(outputws), SummationType='SumInQ', ReductionType='DivergentBeam', AnalysisMode='MultiDetectorAnalysis', ProcessingInstructions=processing, FirstTransmissionRun="", CorrectionAlgorithm='None', MomentumTransferStep=resolution,  OutputWorkspaceBinned='QSD'+str(outputws)+'_binned', OutputWorkspace='QSD'+str(outputws)+'_refQ', OutputWorkspaceWavelength='QSD'+str(outputws)+'_refL',StrictSpectrumChecking=False, WavelengthMax=13.5,Diagnostics=diagnostics)
    #Q summed for bent sample
    ReflectometryReductionOneAuto(InputWorkspace=str(outputws), SummationType='SumInQ', ReductionType='NonFlatSample', AnalysisMode='MultiDetectorAnalysis', ProcessingInstructions=processing, FirstTransmissionRun="", CorrectionAlgorithm='None', MomentumTransferStep=resolution,  OutputWorkspaceBinned='QSB'+str(outputws)+'_binned', OutputWorkspace='QSB'+str(outputws)+'_refQ', OutputWorkspaceWavelength='QSB'+str(outputws)+'_refL',StrictSpectrumChecking=False, WavelengthMax=13.5,Diagnostics=diagnostics)
    plotResultsQ(outputws,"Q_"+outputws,xmin,xmax,ymin,ymax)
    
#Now create a routine to put data into any given pixel. Choose to have a step at exactly 6A
#Offspec IDF--> Source at -23.83m detector at 3.62m, total=27.45m
#Speed of 6A neutron is 659.3390 ms-1
#ToF=0.0416326s = 41632.6 us closest in data is 41635us = [2081]
def pixelwriter(pixel,wksp):
    print "Updating pixel with index " + str(pixel)
    a=mtd[str(wksp)]
    for time in range(0,2082):
        a.dataY(pixel)[time]=1
    for time in range(2082,5000):
        a.dataY(pixel)[time]=10
    for time in range(0,5000):
        for monitor in range(0,3):
            a.dataY(monitor)[time]=1000000

perfect_data = Load("perfect_zero_data.nxs")
pixelwriter(pixelindex,perfect_data)

# Do the reduction for the given range
start=pix_start
end=pix_end
instructions=str(start) + "-" + str(end)
ws_name="perfect_"+str(start+1) + "-" + str(end+1)
CloneWorkspace(InputWorkspace="perfect_data",OutputWorkspace=ws_name)
perfectreduce(outputws=ws_name, processing=instructions,resolution=0.001,xmin=0.0142,xmax=0.0148,ymin=7e7,ymax=1.3e10)
```

1. This screenshot shows the expected plot:

![simple_test_plot](https://user-images.githubusercontent.com/12895056/28789874-00191b20-761f-11e7-8ed5-0016eedba04f.png)

2. This screenshot shows the additional interim workspaces that should be output when diagnostics are enabled (named `perfect_390-419_...`):

![simple_test_workspaces](https://user-images.githubusercontent.com/12895056/28789884-02b055f6-761f-11e7-9089-e0df6afbb993.png)


---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
